Overridden abstract methods using generics are not longer called for partial mock (working with Mockito 1.10.19)

Hi,

We migrated to Mockito 2.x not so long ago and we came across a test that used to works back in Mockito 1 (1.10.19) and no longer works with the latest 2.x version (2.5.6).

Here is the SSCCE to replicate the issue:

```
package com.some.company;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;

import javax.swing.JButton;

import org.junit.Test;

public class Mockito2Issue
{
	public abstract class SomeAbstractClass<T>
	{
		protected abstract String getRealValue(T value);

		public String getValue(T value)
		{
			return getRealValue(value);
		}
	}

	public class SomeConcreteClass<T extends java.awt.Component> extends SomeAbstractClass<T>
	{
		@Override
		protected String getRealValue(T value)
		{
			return "RealValue";
		}
	}

	@Test
	public void testBug() throws IOException, InstantiationException, IllegalAccessException, InvocationTargetException
	{
		SomeConcreteClass<java.awt.Component> testBug = spy(new SomeConcreteClass<java.awt.Component>());
		assertEquals("RealValue", testBug.getValue(new JButton()));
	}
}
```

**Before (1.10.19):**
Used to works

**After (2.5.6)**
```
java.lang.AssertionError: expected:<RealValue> but was:<null>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at com.some.company.Mockito2Issue.testBug(Mockito2Issue.java:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:44)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:41)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
	at org.junit.runners.BlockJUnit4ClassRunner.runNotIgnored(BlockJUnit4ClassRunner.java:79)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:71)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:49)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

We performed a few side tests and here are the results:
- if we change the method `getRealValue` in class `SomeAbstractClass` to be non-abstract but rather throw an exception, the test pass:
```
		protected String getRealValue(T value)
		{
			throw new UnsupportedOperationException();
		}
```
- if we no longer use generic on the method (e.g. using Object rather than T), the test pass.

Can you take a look at this?


Environment details:

Component | Version
--------- | ------
OS | Ubuntu 16.10
Java | Oracle Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Eclipse | Neon.2
JUnit | 4.8.2
Mockito | 1.10.19 (before), 2.5.6 (after)

--- a/build.gradle
+++ b/build.gradle
@@ -45,8 +45,8 @@ apply from: 'gradle/root/version.gradle'
 apply from: "gradle/java-library.gradle"
 
 dependencies {
-    compile 'net.bytebuddy:byte-buddy:1.6.2'
-    compile 'net.bytebuddy:byte-buddy-agent:1.6.2'
+    compile 'net.bytebuddy:byte-buddy:1.6.4'
+    compile 'net.bytebuddy:byte-buddy-agent:1.6.4'
 
     provided "junit:junit:4.12", "org.hamcrest:hamcrest-core:1.3"
     compile "org.objenesis:objenesis:2.5"

--- a/src/test/java/org/mockitousage/constructor/CreatingMocksWithConstructorTest.java
+++ b/src/test/java/org/mockitousage/constructor/CreatingMocksWithConstructorTest.java
@@ -127,4 +127,27 @@ public class CreatingMocksWithConstructorTest extends TestBase {
         List list = mock(List.class, withSettings().defaultAnswer(CALLS_REAL_METHODS));
         assertNull(list.get(1));
     }
+
+    @Test
+    public void handles_bridge_methods_correctly() {
+        SomeConcreteClass<Integer> testBug = spy(new SomeConcreteClass<Integer>());
+        assertEquals("value", testBug.getValue(0));
+    }
+
+    public abstract class SomeAbstractClass<T> {
+
+        protected abstract String getRealValue(T value);
+
+        public String getValue(T value) {
+            return getRealValue(value);
+        }
+    }
+
+    public class SomeConcreteClass<T extends Number> extends SomeAbstractClass<T> {
+
+        @Override
+        protected String getRealValue(T value) {
+            return "value";
+        }
+    }
 }

--- a/subprojects/android/android.gradle
+++ b/subprojects/android/android.gradle
@@ -11,7 +11,7 @@ apply from: "$rootDir/gradle/java-library.gradle"
 
 dependencies {
     compile project.rootProject
-    compile "net.bytebuddy:byte-buddy-android:1.6.2"
+    compile "net.bytebuddy:byte-buddy-android:1.6.4"
 }
 
 apply from: "$rootDir/gradle/publishable-java-library.gradle"

